### PR TITLE
[Merged by Bors] - TY-2722 method on engine to search by topic

### PR DESCRIPTION
**Pull request by finiteprods**
previously #330 #339

**Summary**

adds a method to the engine for searching by topic. also:

- adds ffi binding for above
- corresponding changes on the dart side to accommodate new engine method
- search manager calls the new method when search requested is by topic

### CODE REVIEW

File: discovery_engine/lib/src/domain/models/active_search.dart [hunk 1]
```diff
@@ -26,7 +26,7 @@ part 'active_search.g.dart';
 class ActiveSearch with _$ActiveSearch {
   @HiveType(typeId: searchTypeId)
   const factory ActiveSearch({
-    @HiveField(0) required String queryTerm,
+    @HiveField(0) required String searchTerm,
```
Comment (Robert-Steiner) [inline discovery_engine/lib/src/domain/models/active_search.dart:29]: ❔I know too little about the hive db but this change doesn't require any mirgration, does it?

File: discovery_engine/lib/src/domain/search_manager.dart [hunk 2]
```diff
@@ -100,7 +113,7 @@ class SearchManager {
     await searchClosed();
 
     final search = ActiveSearch(
-      queryTerm: queryTerm,
+      searchTerm: queryTerm,
```
Comment (Robert-Steiner) [inline discovery_engine/lib/src/domain/search_manager.dart:116]: ❔Should we also rename `queryTerm` to `searchTerm`?

File: discovery_engine_core/core/src/engine.rs [hunk 3]
```diff
@@ -457,6 +464,66 @@ where
...
+        }
+
+        let mut errors = Vec::new();
+        let mut articles = Vec::new();
+
+        let markets = self.config.markets.read().await;
+        let scaled_page_size = page_size as usize / markets.len() + 1;
+        let excluded_sources = self.config.excluded_sources.read().await.clone();
+        for market in markets.iter() {
+            let news_query = HeadlinesQuery {
```
Comment (Robert-Steiner) [inline discovery_engine_core/core/src/engine.rs:485]:
```suggestion
            let headlines_query = HeadlinesQuery {
```

File: discovery_engine_core/core/src/engine.rs [hunk 4]
```diff
@@ -457,6 +464,66 @@ where
...
     }
 
+    /// Perform an active search by topic.
+    pub async fn search_by_topic(
+        &mut self,
+        topic: &str,
+        page: u32,
+        page_size: u32,
+    ) -> Result<Vec<Document>, Error> {
+        if topic.trim().is_empty() {
```
Comment (Robert-Steiner) [inline discovery_engine_core/core/src/engine.rs:474]: 🌱 Except for the query, both methods ( `active_search` and `search_by_topic`) share the same code. Maybe it makes sense to refactor the code a bit in a follow-up future.

**Review by Robert-Steiner — APPROVED** [review]
🚀

File: discovery_engine/lib/src/domain/models/active_search.dart (hunk 1 above)
Comment (finiteprods) [reply discovery_engine/lib/src/domain/models/active_search.dart:29]: good question. according to the hive docs, renaming (while keeping the `HiveField` number the same) should be fine.

Comment (biernacki) [reply discovery_engine/lib/src/domain/models/active_search.dart:29]: yes, it should be fine, field name is not important as long as it uses the same `HiveField` identifier.

**Comment by finiteprods** [comment]
bors merge

**Pull request closed without merging**